### PR TITLE
Pnudev 4

### LINKED ISSUE #4: Create, read, update, delete functionality for partners
1. Create domain model
2. Create migration
3. PartnerController
  - GET /partners
  - GET POST /partners/new
  - GET POST /partners/edit/{partnerId}
  - POST /partners/delete/{partnerId}

### PR DESCRIPTION
**Pull request by MrZenik**

### CODE REVIEW

File: src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Controller
+@RequestMapping("/partners")
+public class PartnerController {
+    private PartnerServiceImpl partnerService;
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java]: Add new line before field

Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java]: Change this field to abstract PartnerService (not impl)

File: src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        partnerService.deletePartnerById(id);
+        return "redirect:/partners";
+    }
+
+    @GetMapping("/edit/{id}")
+    public String editPartner(@PathVariable Long id, Model model){
+
+        Optional<Partner> optionalPartner = partnerService.findPartnerById(id);
+        Partner partner = optionalPartner.orElseThrow();
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java]: PartnerService should return Partner instance, if Partner not found, throw exception in service

File: src/main/java/com/yevhenberladyniuk/debts/domain/Partner.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    private Long userId;
+
+    private String firstName;
+
+    private String lastName;
+
+    private double debt;
+
+    private LocalDateTime date;
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/domain/Partner.java]: Rename field to `updatedAt`

File: src/main/java/com/yevhenberladyniuk/debts/domain/Partner.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.data.annotation.Id;
+
+import java.time.LocalDateTime;
+
+@Data
+@Builder(toBuilder = true)
+@NoArgsConstructor
+@AllArgsConstructor
+public class Partner {
+
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/domain/Partner.java:16]: Add boolean field `active`, and when delete, change it to false

File: src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.Id;
+
+import javax.validation.constraints.NotBlank;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PartnerDto {
+
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java:15]: Add field `debt` which represents total debt
And `updatedAt` will be fine here too

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import com.yevhenberladyniuk.debts.domain.Partner;
+import com.yevhenberladyniuk.debts.dto.CreatePartnerForm;
+import com.yevhenberladyniuk.debts.dto.PartnerDto;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface PartnerService {
+
+    void create(CreatePartnerForm createPartnerForm, Long userId);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: This method should receive `createPartnerForm` and `user`, not `userId`

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    List<Partner> findAllByUserId(Long userId);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: `user`, not `userId`
and rename to `findAll`

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    void deletePartnerById(Long id);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: should receive `user` also, and in logic you should verify whether partner belongs to authenticated user

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    Optional<Partner> findPartnerById(Long id);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: method should receive `user`, cause now any user can receive any partner by id
Each user is allowed only to find partners that they have created

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+    void updatePartnerById(Long id, PartnerDto partnerDto);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: same as with delete

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    public Optional<Partner> findPartnerById(Long id) {
+        return partnerRepository.findById(id);
+    }
+
+    @Override
+    public void updatePartnerById(Long id, PartnerDto partnerDto) {
+
+        Optional<Partner> optionalPartner = findPartnerById(id);
+        Partner partner = optionalPartner.orElseThrow();
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java]: `Partner partner = optionalPartner.orElseThrow( () -> new RuntimeException("Partner not found"));`

File: src/main/resources/db/migration/V1_1__CreatePartner.sql [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    id         serial       not null,
+    user_id    serial       not null,
+    first_name varchar(255) not null,
+    last_name  varchar(255) not null,
+    debt       decimal      not null,
+    date       timestamp    not null,
+
+    primary key (id),
+    foreign key(user_id) references "user"(id)
+);
```
Comment (volodymyrvynnyk) [inline src/main/resources/db/migration/V1_1__CreatePartner.sql]: Rename file to V2__CreatePartner.sql

**Review by volodymyrvynnyk — CHANGES_REQUESTED** [review]
Main problem - now any user is allowed to delete/update/find any partner
In each method you have to check whether `partner.getUserId == user.getId`, and get data from repository with `ByUserId` key

File: src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java [hunk 13]
```diff
@@ -12,14 +12,14 @@
...
 
 @Controller
 @RequestMapping("/partners")
 public class PartnerController {
-    private PartnerServiceImpl partnerService;
+
+    private PartnerService partnerService;
 
     @Autowired
     public PartnerController(PartnerServiceImpl partnerService) {
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/controller/PartnerController.java]: Change attribute to PartnerService

File: src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java [hunk 14]
```diff
@@ -21,4 +22,9 @@
     @NotBlank
     private String lastName;
 
+    @NotBlank
+    private double debt;
+
+    @NotBlank
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java]: `@NotBlank` is used for `String`
By the way, there's no needs of these constraints in DTO that is returned from server

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java [hunk 15]
```diff
@@ -1,21 +1,21 @@
...
+    List<Partner> findAll(User user);
 
-    void deletePartnerById(Long id);
+    void deletePartnerById(Long id, User user);
 
-    Optional<Partner> findPartnerById(Long id);
+    Partner findPartnerById(Long id, User user);
 
-    void updatePartnerById(Long id, PartnerDto partnerDto);
+    void updatePartnerById(Long id, PartnerDto partnerDto, User user);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerService.java]: You can skip the word 'Partner' in method signature, not necessary, but looks better

File: src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java (hunk 14 above)
Comment (volodymyrvynnyk) [reply src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java]: Is this class used somewhere?

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java [hunk 16]
```diff
@@ -22,42 +23,52 @@ public PartnerServiceImpl(PartnerRepository partnerRepository) {
...
+        partner.setActive(false);
+        partnerRepository.save(partner);
     }
 
     @Override
-    public Optional<Partner> findPartnerById(Long id) {
-        return partnerRepository.findById(id);
+    public Partner findPartnerById(Long id, User user) {
+
+        Optional<Partner> optionalPartner = partnerRepository.findByIdAndActive(id, true);
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java]: Add method `findByIdAndActiveAndUsedId` to `repository` and use it

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java [hunk 17]
```diff
@@ -22,42 +23,52 @@ public PartnerServiceImpl(PartnerRepository partnerRepository) {
...
+        Partner partner = optionalPartner.orElseThrow( () -> new RuntimeException("Partner not found"));
+
+        if(partner.getUserId().equals(user.getId())){
+            return partner;
+        }
+
+        throw new RuntimeException("Partner not found");
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java]: When you add `findByIdAndActiveAndUsedId` method, no needs if this userId validation

**Review by volodymyrvynnyk — CHANGES_REQUESTED** [review]
Few comments to change the code
Almost ready to merge! Good job

File: src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java (hunk 14 above)
Comment (MrZenik) [reply src/main/java/com/yevhenberladyniuk/debts/dto/PartnerDto.java]: yes, inside the ```PartnerService.updateById``` method

File: src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java [hunk 18]
```diff
@@ -43,29 +43,23 @@ public void create(CreatePartnerForm createPartnerForm, User user) {
...
 
     @Override
-    public void updatePartnerById(Long id, PartnerDto partnerDto, User user) {
+    public void updateById(Long id, PartnerDto partnerDto, User user) {
 
-        Partner partner = findPartnerById(id, user);
+        Partner partner = findById(id, user);
         partner.setLastName(partnerDto.getLastName());
         partner.setFirstName(partnerDto.getFirstName());
         partner.setUpdatedAt(LocalDateTime.now());
```
Comment (volodymyrvynnyk) [inline src/main/java/com/yevhenberladyniuk/debts/service/PartnerServiceImpl.java]: set debt from dto

**Review by volodymyrvynnyk — APPROVED** [review]
One comment and can merge

**Pull request merged**